### PR TITLE
A more minimal Ignition to Gazebo server rename effort

### PR DESCRIPTION
**Pull request by nkoenig**
Signed-off-by: Nate Koenig <nate@openrobotics.org>

# 🎉 New feature

* Part of https://github.com/gazebo-tooling/release-tools/issues/698

## Summary

For Citadel and Fortress users I'd like to achieve the following goals:

1. Assets that reference fuel.ignitionrobotics.org continue to work
2. Assets the reference fuel.gazebosim.org work
3. Assets installed in ~/.ignition/fuel

## Test it

1. Clear your `~/.ignition/fuel` directory.
2. Run this world file, which uses a `fuel.gazebosim.org` URI. The included model depends on another model whose URI uses `fuel.ignitionrobotics.org`.
  ```
<?xml version="1.0" ?>
<sdf version="1.6">
  <world name="shapes">
    <plugin
      filename="ignition-gazebo-physics-system"
      name="ignition::gazebo::systems::Physics">
    </plugin>
    <plugin
      filename="ignition-gazebo-user-commands-system"
      name="ignition::gazebo::systems::UserCommands">
    </plugin>
    <plugin
      filename="ignition-gazebo-scene-broadcaster-system"
      name="ignition::gazebo::systems::SceneBroadcaster">
    </plugin>

    <scene>
    <ambient>1.0 1.0 1.0</ambient>
    <background>0.8 0.8 0.8</background>
    </scene>

    <include>
      <uri>https://fuel.gazebosim.org/1.0/OpenRobotics/models/Tunnel Tile 4</uri>
    </include>
  </world>
</sdf>
  ```
4. You should see `Tunnel Tile 4` in `~/.ignition/fuel/fuel.gazebosim.org/openrobotics`, and `tunnel tile 1` in `~/.ignition/fuel/fuel.ignitionrobotics.org/openrobotics`.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/FuelClient_TEST.cc [hunk 1]
```diff
@@ -538,12 +538,14 @@ TEST_F(FuelClientTest, DownloadModel)
     EXPECT_EQ(ResultType::FETCH_ALREADY_EXISTS, res5.Type());
   }
 
-  // Download model with a dependency specified within its `model.config`
+  // Download model with a dependency specified within its `model.config`.
+  // The dependency points to fuel.gazebosim.org.
```
Comment (chapulina) [inline src/FuelClient_TEST.cc:542]: Is it valuable to also test the other way around, i.e. a `ignitionrobotics` URL that depends on a `gazebosim` URL?

File: src/FuelClient.cc [hunk 2]
```diff
@@ -572,12 +572,16 @@ Result FuelClient::DownloadModel(const ModelIdentifier &_id,
...
     return res;
 
-  for (auto dep : dependencies)
+  for (ModelIdentifier dep : dependencies)
   {
-    auto dep_res = this->DownloadModel(dep, _headers);
+    // Download dependency if not in the local cache
+    if (!this->dataPtr->cache->MatchingModel(dep))
+    {
+      auto dep_res = this->DownloadModel(dep, _headers);
```
Comment (chapulina) [inline src/FuelClient.cc:580]: Style nit (I know it's pre-existing though)

```suggestion
      auto depRes = this->DownloadModel(dep, _headers);
```

File: src/ClientConfig.cc [hunk 3]
```diff
@@ -44,6 +44,12 @@ class ignition::fuel_tools::ClientConfigPrivate
                 homePath, ".ignition", "fuel");
 
             this->servers.push_back(ServerConfig());
+
+            // Add in fuel.gazebosim.org as another default server config.
+            ServerConfig gzServerConfig;
+            gzServerConfig.SetUrl(common::URI("https://fuel.gazebosim.org"));
+            gzServerConfig.SetVersion("1.0");
+            this->servers.push_back(gzServerConfig);
```
Comment (chapulina) [inline src/ClientConfig.cc:52]: Just for completeness, is there an advantage to hardcoding it here as opposed to there

https://github.com/gazebosim/gz-fuel-tools/blob/b3c253c0177192e6a64683d9ccd1559de3a521e6/conf/config.yaml#L3-L6

In theory, I think that hardcoding it here makes it difficult, if not impossible for users of other servers to disable `gazebosim.org`. Not sure if any users at all care about that, but it's a use case supported to an extent.

File: src/ClientConfig_TEST.cc [hunk 4]
```diff
@@ -76,7 +76,7 @@ std::string cachePath()
 TEST(ClientConfig, InitiallyDefaultServers)
 {
   ClientConfig config;
-  EXPECT_EQ(1u, config.Servers().size());
+  EXPECT_EQ(2u, config.Servers().size());
```
Comment (chapulina) [inline src/ClientConfig_TEST.cc:79]: We should add a note to the migration guide, since this is a slight behaviour change.

**Review by chapulina — APPROVED** [review]
Works as advertised

```
$ tree ~/.ignition/fuel/ -L 4
/home/chapulina/.ignition/fuel/
├── fuel.gazebosim.org
│   └── openrobotics
│       └── models
│           └── tunnel tile 4
└── fuel.ignitionrobotics.org
    └── openrobotics
        └── models
            └── tunnel tile 1
```

I just have some minor comments below, but nothing blocking.

File: src/ClientConfig.cc (hunk 3 above)
Comment (nkoenig) [reply src/ClientConfig.cc:52]: There are a couple problems associated with the yaml config file.

1. I don't think it's possible to use the yaml config file when running `ign gazebo`.
2. I believe the config file is only used with `ign fuel` if the `--config` option is specified.
3. If either `fuel.ignitionrobotics.org` or `fuel.gazebosim.org` are missing in the ClientConfig, then worlds/models that use both URIs will not load.

I don't see an easy way around this.

File: src/ClientConfig_TEST.cc (hunk 4 above)
Comment (nkoenig) [reply src/ClientConfig_TEST.cc:79]: Added in [352f42d](https://github.com/gazebosim/gz-fuel-tools/pull/263/commits/352f42d3ea70743024145517f37203f23b68259f)

File: src/FuelClient.cc (hunk 2 above)
Comment (nkoenig) [reply src/FuelClient.cc:580]: Changed in [352f42d](https://github.com/gazebosim/gz-fuel-tools/pull/263/commits/352f42d3ea70743024145517f37203f23b68259f)

File: src/FuelClient_TEST.cc (hunk 1 above)
Comment (nkoenig) [reply src/FuelClient_TEST.cc:542]: The reverse also works. I've just added the ability to use both `fuel.ignitionrobotics.org` and 'fuel.gazebosim.org' at the same time. As long as both server configurations are present, then it'll work.

**Pull request merged**